### PR TITLE
Adds module:config:status command which checks if the module config i…

### PR DESCRIPTION
**Pull request by hostep**
…n the app/etc/config.php file is correct or not.

### Description (*)
This adds a new command to `bin/magento`: `module:config:status`.
This command checks if the modules in the file `app/etc/config.php` are correct or not.

I've created this command since I'm getting a bit tired of seeing incorrect `app/etc/config.php` files being added to our git repo's whenever somebody adds a new module using composer and forgets to run `bin/magento setup:upgrade`, or when some branch get merged which has merge conflicts in the `config.php` file and the conflict gets solved incorrectly.
Having this new command, would allow us to have a pre-commit githook which can check for this problem and then tell the developer to fix the `config.php` file before committing.

Feel free to let me know if the command name or description needs to be changed. And also if the output messages need to be changed.

### Fixed Issues (if relevant)
None that I could find.

### Manual testing scenarios (*)
1. Have a correct Magento installation
2. Run `bin/magento module:config:status`, it should say: `The modules configuration is up to date.`
3. Edit the file `app/etc/config.php` and remove or add a module. Or just switch the sort order of 2 modules.
4. Run `bin/magento module:config:status`, it should say: `The modules configuration in the 'app/etc/config.php' file is outdated. Run 'setup:upgrade' to fix it.`
5. Run `bin/magento setup:upgrade`
6. Run `bin/magento module:config:status`, it should say: `The modules configuration is up to date.`

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: setup/src/Magento/Setup/Console/Command/ModuleConfigStatusCommand.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            $correctModuleConfig = $installer->getModulesConfig();
+
+            if ($currentModuleConfig !== $correctModuleConfig) {
+                throw new \Exception(
+                    'The current module config in the \'app/etc/config.php\' file is incorrect. Run \'setup:upgrade\' to fix it.'
+                );
+            }
+
+            $output->writeln(
+                '<info>Everything seems to be in order.</info>'
```
Comment (dmytro-ch) [inline setup/src/Magento/Setup/Console/Command/ModuleConfigStatusCommand.php]: Could you please change the success message to something more certain? :)
e.g. `All modules configurations are up to date.`

File: setup/src/Magento/Setup/Console/Command/ModuleConfigStatusCommand.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            );
+        } catch (\Exception $e) {
+            $output->writeln('<error>' . $e->getMessage() . '</error>');
+
+            return \Magento\Framework\Console\Cli::RETURN_FAILURE;
```
Comment (dmytro-ch) [inline setup/src/Magento/Setup/Console/Command/ModuleConfigStatusCommand.php]: Please move the `Magento\Framework\Console\Cli` class name to the imports as well.
```
use Magento\Framework\Console\Cli;
```
```
return Cli::RETURN_FAILURE;
```

File: setup/src/Magento/Setup/Model/Installer.php [hunk 3]
```diff
@@ -1196,6 +1196,17 @@ public function updateModulesSequence($keepGeneratedFiles = false)
         $this->createModulesConfig([]);
     }
 
+    /**
+     * Get the modules config as Magento sees it
+     *
+     * @return array
+     * @throws \LogicException
+     */
+    public function getModulesConfig()
```
Comment (dmytro-ch) [inline setup/src/Magento/Setup/Model/Installer.php:1209]: Could you please avoid the creation of new public methods.
I would recommend adding this implementation as a private method of `Magento\Setup\Console\Command\ ModuleConfigStatusCommand` class.

**Review by dmytro-ch — CHANGES_REQUESTED** [review]

File: setup/src/Magento/Setup/Console/Command/ModuleConfigStatusCommand.php (hunk 1 above)
Comment (hostep) [reply setup/src/Magento/Setup/Console/Command/ModuleConfigStatusCommand.php]: Thanks for the review, I'll try to address your comments later today!

File: setup/src/Magento/Setup/Model/Installer.php (hunk 3 above)
Comment (hostep) [reply setup/src/Magento/Setup/Model/Installer.php:1209]: The problem is that the `createModulesConfig` function is private, so I need to expose it somehow without copying it over completely.
According to the [semver checker result](https://results.solutionascode.com/results/magento/magento2/pull/22733/ea76f902-95f0-491c-82a7-5922f485cf57/10891/SemanticVersionChecker/report-magento2.html) it is ok to add a public function to this class.
What do you think?

Comment (dmytro-ch) [reply setup/src/Magento/Setup/Model/Installer.php:1209]: > The problem is that the createModulesConfig function is private

You are right, I didn't notice this point.
So let's keep this method.
Thank you.

**Comment by hostep** [comment]
@dmytro-ch: I've force pushed the requested changes.
I changed some other output messages and the description of the command as well, so they are more consistent with each other.

**Review by dmytro-ch — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch, thank you for the review.
[ENGCOM-5080](https://jira.corp.magento.com/browse/ENGCOM-5080) has been created to process this Pull Request

**Comment by dmytro-ch** [comment]
@hostep thank you!

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Comment by hostep** [comment]
Thanks for the cleanup @nmalevanec! I'm curious which command you ran to detect these problems, so the next time I can try to fix these myself when creating a new PR.

**Comment by nmalevanec** [comment]
Hi, @hostep. You can just run dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php test against modified php files.

**Comment by sivaschenko** [comment]
@piotrekkaminski can you please take a look at this feature and provide your feedback

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by hostep** [comment]
@sivaschenko: is it ok if I still update my PR? There are some missing `declare(strict_types=1);` lines, it seems like I forgot to add these. And should I add these to new classes only, or also to other classes touched by this PR?

I'll then also do some git rebasing of this PR on top of the latest 2.3-develop, maybe that helps with these tests which seem to be not running somehow.

I also still don't understand how to execute static tests on my local, the `dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php` file just contains a class and no executable code, so I'm a bit confused about the statement above.

Thanks!

**Comment by hostep** [comment]
Rebased PR on latest 2.3-develop branch and added `declare(strict_types=1);` to the 2 new classes.

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by hostep** [comment]
Force pushed again, which will hopefully satisfy the static tests (I've amended @nmalevanec's commit for this, as not to loose his authorship of the other fixes).

It would still be great if somebody can tell me how to execute these static tests on my local environment, so I don't have to wait 2 days for the results and can just fix them in one go, thanks! 🙂

**Pull request merged**